# 1. 데이터 가져오기 

In [169]:
import pandas as pd
import numpy as np
import re

In [170]:
review4us = pd.read_csv("자사리뷰.csv")
review4rivals = pd.read_csv("경쟁사리뷰.csv")

review4us.tail(100)

Unnamed: 0 title rating        date      time  view  \
27119       27119    만족     5점  2018-08-26  02:21:43    17   
27120       27120    보통     3점  2018-08-25  02:34:32    18   
27121       27121    만족     5점  2018-08-25  02:23:30    20   
27122       27122    만족     5점  2018-08-25  02:23:30    17   
27123       27123    보통     3점  2018-08-24  02:49:49    16   
...           ...   ...    ...         ...       ...   ...   
27214       27214    만족     5점  2018-07-11  02:24:45     7   
27215       27215    만족     5점  2018-07-07  02:41:20     8   
27216       27216    만족     5점  2018-07-02  02:38:11    33   
27217       27217    만족     5점  2018-06-27  02:56:11    34   
27218       27218    만족     5점  2018-06-26  02:29:45    35   

                                             text  naver  photo  
27119                                       만족합니다      1      0  
27120                    먹은지 일주일이지만, 아직은 잘 모르겠어요!      1      0  
27121                          더 먹어봐야 알겟지만 좋은거같아요      1      0  
27122                              좋네요오오오ㅎㅎㅇㅋ오 ㅎㅎ      1      0  
27123                      계속 먹어보면 알겠죠 아직은 별효과업근긋      1      0  
...                                           ...    ...    ...  
27214                                개별포장 편하고 좋네요      1      0  
27215                         여자친구님께서 매우 만족하셨습니다.      1      0  
27216                          좋은구성이네요 효과봤으면 좋겠네요      1      0  
27217                                      이제받았어요      1      0  
27218  배송이 좀 늦었어요...ㅠㅠ\n원하는 날부터 시작은 못했지만 기대해봅니다ㅎㅎ      1      0  

[100 rows x 9 columns]

In [171]:
review4rivals.tail(100)

Unnamed: 0       date                                 item  \
36297       36297  17.08.03.  ★상품명 반드시 확인必★: 01 온가족 유산균 락토핏 골드 1통   
36298       36298  17.08.03.  ★상품명 반드시 확인必★: 01 온가족 유산균 락토핏 골드 1통   
36299       36299  17.08.02.  ★상품명 반드시 확인必★: 01 온가족 유산균 락토핏 골드 1통   
36300       36300  17.08.02.  ★상품명 반드시 확인必★: 01 온가족 유산균 락토핏 골드 1통   
36301       36301  17.08.02.  ★상품명 반드시 확인必★: 01 온가족 유산균 락토핏 골드 1통   
...           ...        ...                                  ...   
36392       36392  17.05.22.              선택1: 락토핏 코스트코 생유산균19 1통   
36393       36393  17.05.20.                  선택1: 락토핏 생유산균 골드 1통   
36394       36394  17.05.19.                  선택1: 락토핏 생유산균 골드 1통   
36395       36395  17.05.19.              선택1: 락토핏 코스트코 생유산균19 1통   
36396       36396  17.05.18.                  선택1: 락토핏 생유산균 골드 1통   

                                                    text  rating user_name  \
36297  직접 먹어본 후 부모님께 사드렸어요 변비뿐만 아니라 면역력에도 좋다고 해서 제가 직...       5  mink****   
36298                    제가 직접 먹어본 후 효과 좋아서 부모님께도 사드렸어요~       5  mink****   
36299                                            빨리와서좋아용       3  rlat****   
36300                                좋아요 배송도 빠르고 늘 먹고있엉ㅅ       5  aroh****   
36301  좋습니다ㅎ 벌써 3번째 구입이네요ㅎㅎ\n이거먹고는 매일 화장실가고있어요 매일 한개씩...       5  alsd****   
...                                                  ...     ...       ...   
36392                       잘먹을게요~^^ 배송이 빨라서 너무 만족합니당ㅎㅎ        5  ayh6****   
36393      좋은 상품감사합니다\n빠른배송에\n잘챙겨먹으려고요\n또 구매할게요\n많이파세요^^       5  8495****   
36394                                             다들잘먹네요       5  ll97****   
36395                                       빠른 배송 감사드려요        5  youn****   
36396                                             쪼아영 ㅎㅋ       5  zhen****   

       photo  one_month  repurchase  pre_reviewed  
36297      0          0           0             0  
36298      0          0           0             0  
36299      0          0           0             0  
36300      0          0           0             0  
36301      0          0           0             0  
...      ...        ...         ...           ...  
36392      0          0           0             0  
36393      0          0           0             0  
36394      0          0           0             0  
36395      0          0           0             0  
36396      0          0           0             0  

[100 rows x 10 columns]

In [172]:
#unnamed는 필요없는 열이므로 제거한다
del review4rivals['Unnamed: 0']
del review4us['Unnamed: 0']

# Unnamed 열이 사라졌음을 볼 수 있다. 
print(review4rivals.columns, review4us.columns)

Index(['date', 'item', 'text', 'rating', 'user_name', 'photo', 'one_month',
       'repurchase', 'pre_reviewed'],
      dtype='object') Index(['title', 'rating', 'date', 'time', 'view', 'text', 'naver', 'photo'], dtype='object')


# 2. 분석 전 데이터 전처리 
* review4rivlas 의 item 종류에 각종 특수문자, 의미없는 숫자 제거하기. 
- https://regexr.com/ 이 사이트를 이용하면 정규표현식을 그나마 손쉽게 try out 할 수 있다. 
* text colum에 있는 특수문자들 제거하기. 
* unnamed column은 뭐지? 

### 2-1. 경쟁사 제품 리뷰 데이터 전처리

In [173]:
# item 종류에 있는 각종 특수문자와 의미없는 문자 제거하기 : ': 세미콜론 기준 왼쪽에 있는 글자는 다 없애자.  
def modify1(item):
    x=re.sub(r'[^.]*:([^ ]*)', '', str(item))     ## 이거 알아내는데 5시간 걸렸다.. 실환가?
    return(x)   

review4rivals['item'] = review4rivals['item'].apply(modify1)

def modify2(item):                    # 특수문자 제거 
    x=re.sub(r'[^\s\w]','',str(item))               
    return(x)
review4rivals['item'] = review4rivals['item'].apply(modify2)

def modify3(item):                   #락토핏 앞에 거슬렸던 2, 9, 3과 같은 숫자들이 사라진다.  그러나 완벽하게 사라진 것은 아니다. 
    x=re.sub('\d\d*.?(락토핏)','락토핏',str(item))
    return(x)
review4rivals['item'] = review4rivals['item'].apply(modify3)

def modify4(item):
    x=re.sub('\d\d?\d?[가-힣]','',str(item))     # 몇통, 몇포와 같은 수량정보를 제거한다.
    return(x)

In [174]:
print(review4rivals['item'].head(100))
print(review4rivals.item.tail(1000))

0         락토핏 생유산균 골드3통150포
1         락토핏 생유산균 골드3통150포
2         락토핏 생유산균 골드3통150포
3             락토핏 생유산균 키즈1통
4         락토핏 생유산균 골드5통250포
              ...          
95         락토핏 생유산균 코어1통60포
96         락토핏 생유산균 골드1통50포
97        락토핏 생유산균 골드3통150포
98         락토핏 생유산균 골드1통50포
99     락토핏 생유산균 맘스1통 임산부 추천
Name: item, Length: 100, dtype: object
35397       03 청소년유산균 락토핏  키즈 1통
35398     02 코스트코 장을위한 생유산균19 1통
35399       03 청소년유산균 락토핏  키즈 1통
35400       01 온가족 유산균 락토핏 골드 1통
35401     02 코스트코 장을위한 생유산균19 1통
                  ...           
36392         락토핏 코스트코 생유산균19 1통
36393             락토핏 생유산균 골드 1통
36394             락토핏 생유산균 골드 1통
36395         락토핏 코스트코 생유산균19 1통
36396             락토핏 생유산균 골드 1통
Name: item, Length: 1000, dtype: object


In [175]:
def text_modify1(text):
    x=re.sub(r'[^가-힣|\d|\s]', '', str(text))     ## 사용자 리뷰에 있는 특수문자와 자, 모음 제거. 숫자는 살아있음. 
    return(x)   

review4rivals['text'] = review4rivals['text'].apply(text_modify1)

In [176]:
review4rivals

date                 item  \
0      21.01.22.    락토핏 생유산균 골드3통150포   
1      21.01.21.    락토핏 생유산균 골드3통150포   
2      21.01.21.    락토핏 생유산균 골드3통150포   
3      21.01.21.        락토핏 생유산균 키즈1통   
4      21.01.21.    락토핏 생유산균 골드5통250포   
...          ...                  ...   
36392  17.05.22.   락토핏 코스트코 생유산균19 1통   
36393  17.05.20.       락토핏 생유산균 골드 1통   
36394  17.05.19.       락토핏 생유산균 골드 1통   
36395  17.05.19.   락토핏 코스트코 생유산균19 1통   
36396  17.05.18.       락토핏 생유산균 골드 1통   

                                                    text  rating user_name  \
0                                  항상 챙겨먹고 있습니다  재구매 할게요       5  zxc9****   
1      꾸준히 먹어야 효과가 있다고 해서 꾸준히 먹는 중이에요 물 없이도 먹기 너무 편하고...       5  ilm0****   
2      여러 유산균을 복용해보다 처음 종근당으로 사봤는데 일단 가격이 저렴해서 부담감없어좋...       4  ki25****   
3                                    주황색 먹이다가 키즈로 구매해봤어요       5  dudw****   
4                                         상품 배송 모두 만족합니다       5  hara****   
...                                                  ...     ...       ...   
36392                            잘먹을게요 배송이 빨라서 너무 만족합니당        5  ayh6****   
36393        좋은 상품감사합니다\n빠른배송에\n잘챙겨먹으려고요\n또 구매할게요\n많이파세요       5  8495****   
36394                                             다들잘먹네요       5  ll97****   
36395                                       빠른 배송 감사드려요        5  youn****   
36396                                               쪼아영        5  zhen****   

       photo  one_month  repurchase  pre_reviewed  
0          0          1           0             0  
1          0          1           0             1  
2          0          0           0             0  
3          0          0           0             0  
4          0          0           0             0  
...      ...        ...         ...           ...  
36392      0          0           0             0  
36393      0          0           0             0  
36394      0          0           0             0  
36395      0          0           0             0  
36396      0          0           0             0  

[36397 rows x 9 columns]

date, user_name, photo, one_month, repurchase, pre_viewd 열의 값들을 알아보고 이 열들이 필요한지 확인하자. 

In [177]:
print(review4rivals.photo.value_counts())
print(review4rivals.repurchase.value_counts())
print(review4rivals.pre_reviewed.value_counts())

0    27319
1     9078
Name: photo, dtype: int64
0    32071
1     4326
Name: repurchase, dtype: int64
0    32496
1     3901
Name: pre_reviewed, dtype: int64


* photo, one_month, repurchase, pre_reviewed 컬럼 모두 범주과 0과 1뿐이다.
* photo열은 단순히 댓글을 달때 구입한 사진을 올렸는지, 아닌지 여부를 표시한 것 같다. 결론을 도출할때 필요한 자료는 아니다. 

* one_month는 무엇을 의미할까 ?

* repurchase는 재구매를 의미한다. 재구매를 한 사람이 쓴 댓글이 4326건. 첫 구매후 쓴 댓글이 32071건. 
  재구매를 한 사람은 어떤 장점 때문에 재구매를 하게 됐을까? 를 repurchase 열을 찾고 해당 행의 text를 보면 알 수 있을 것이다. 

* pre_reviewed는 무엇을 의미할까?

In [178]:
# photo column, user_name, date 제거 
review4rivals = review4rivals.drop(['photo','user_name','date'],axis=1)

In [179]:
review4rivals

item                                               text  \
0        락토핏 생유산균 골드3통150포                              항상 챙겨먹고 있습니다  재구매 할게요   
1        락토핏 생유산균 골드3통150포  꾸준히 먹어야 효과가 있다고 해서 꾸준히 먹는 중이에요 물 없이도 먹기 너무 편하고...   
2        락토핏 생유산균 골드3통150포  여러 유산균을 복용해보다 처음 종근당으로 사봤는데 일단 가격이 저렴해서 부담감없어좋...   
3            락토핏 생유산균 키즈1통                                주황색 먹이다가 키즈로 구매해봤어요   
4        락토핏 생유산균 골드5통250포                                     상품 배송 모두 만족합니다   
...                    ...                                                ...   
36392   락토핏 코스트코 생유산균19 1통                            잘먹을게요 배송이 빨라서 너무 만족합니당    
36393       락토핏 생유산균 골드 1통        좋은 상품감사합니다\n빠른배송에\n잘챙겨먹으려고요\n또 구매할게요\n많이파세요   
36394       락토핏 생유산균 골드 1통                                             다들잘먹네요   
36395   락토핏 코스트코 생유산균19 1통                                       빠른 배송 감사드려요    
36396       락토핏 생유산균 골드 1통                                               쪼아영    

       rating  one_month  repurchase  pre_reviewed  
0           5          1           0             0  
1           5          1           0             1  
2           4          0           0             0  
3           5          0           0             0  
4           5          0           0             0  
...       ...        ...         ...           ...  
36392       5          0           0             0  
36393       5          0           0             0  
36394       5          0           0             0  
36395       5          0           0             0  
36396       5          0           0             0  

[36397 rows x 6 columns]

### 2-2. 자사 제품리뷰 데이터 전처리

In [180]:
review4us.columns

Index(['title', 'rating', 'date', 'time', 'view', 'text', 'naver', 'photo'], dtype='object')